File: Python3/프로그래머스/2/42626. 더 맵게/더 맵게.py
import heapq

def solution(scoville, K):
    #모든 음식의 스코빌지수를 K이상으로 만들기 위해 섞어야 하는 최소 횟수?
    stack = []
    res = 0
    heapq.heapify(scoville)
    
    while scoville:
        if scoville[0] >= K:
            break

        if len(scoville) < 2:
            return -1
        a = heapq.heappop(scoville)
        b = heapq.heappop(scoville)
        c = a + b * 2
        heapq.heappush(scoville, c)
        res += 1
    
    return res
    
    
    
    
    
    
    
    
    
    
    
#     answer = 0  
#     heapq.heapify(scoville)
    
#     while scoville:
#         if scoville[0] >= K:
#             break
        
#         if len(scoville) < 2:
#             answer = -1
#             break
        
#         a = heapq.heappop(scoville) #가장 안매운 음식
#         b = heapq.heappop(scoville) #두번째로 안매운 음식
        
#         c = a + b*2
#         heapq.heappush(scoville, c)
#         answer += 1
            
#     return answer

File: C++/프로그래머스/2/42626. 더 맵게/더 맵게.cpp
#include <bits/stdc++.h>

using namespace std;
//모든 음식의 스코빌 지수가 K이상이 될 때까지 반복하여 섞음, 가장 낮은 두 개의 음식을 계속해서 섞음, queue, 섞는 횟수 리턴
priority_queue<int> heap;

int solution(vector<int> scoville, int K) {
    int answer;
    for (int s : scoville) {
        heap.push(-1 * s); //최소 힙 만들기
    }
    
    bool available = true;
    int cnt = 0;
    cout << heap.top();
    
    while (-heap.top() < K) {
        if (heap.size() < 2) {
            available = false;
            break;
        }
        
        int new_v = -heap.top();
        heap.pop();
        new_v += (-heap.top() * 2);
        heap.pop();
        
        heap.push(-new_v);
        cnt++;
    }
    
    if (available) return cnt;
    return -1;

}